File: src/main/java/com/sotasan/decompiler/transformers/CFRTransformer.java
package com.sotasan.decompiler.transformers;

import com.sotasan.decompiler.models.FileModel;
import org.benf.cfr.reader.api.CfrDriver;
import org.benf.cfr.reader.api.ClassFileSource;
import org.benf.cfr.reader.api.OutputSinkFactory;
import org.benf.cfr.reader.bytecode.analysis.parse.utils.Pair;
import org.benf.cfr.reader.util.getopt.OptionsImpl;
import org.jetbrains.annotations.NotNull;
import java.util.*;
import java.util.concurrent.CompletableFuture;

public class CFRTransformer implements ITransformer, ClassFileSource, OutputSinkFactory, OutputSinkFactory.Sink<String> {

    private static final Map<String, String> OPTIONS = Map.of(
            OptionsImpl.DECOMPILE_INNER_CLASSES.getName(), String.valueOf(false),
            OptionsImpl.RELINK_CONSTANT_STRINGS.getName(), String.valueOf(false),
            OptionsImpl.REMOVE_INNER_CLASS_SYNTHETICS.getName(), String.valueOf(false),
            OptionsImpl.SHOW_CFR_VERSION.getName(), String.valueOf(false)
    );

    private FileModel fileModel;
    private String output;

    @Override
    public CompletableFuture<String> transform(@NotNull FileModel fileModel) {
        return CompletableFuture.supplyAsync(() -> {
            this.fileModel = fileModel;
            CfrDriver driver = new CfrDriver.Builder().withClassFileSource(this).withOptions(OPTIONS).withOutputSink(this).build();
            driver.analyse(Collections.singletonList(fileModel.getPath()));
            if (output.startsWith("/"))
                output = output.substring(30);
            return output.trim();
        });
    }

    @Override
    public Pair<byte[], String> getClassFileContent(String path) {
        return new Pair<>(fileModel.getBytes(), fileModel.getPath());
    }

    @Override
    public String getPossiblyRenamedPath(String path) {
        return path;
    }

    @Override
    public <T> Sink<T> getSink(SinkType sinkType, SinkClass sinkClass) {
        return (Sink<T>) this;
    }

    @Override
    public void write(String sinkable) {
        output = sinkable;
    }

    @Override public void informAnalysisRelativePathDetail(String usePath, String classFilePath) {}
    @Override public Collection<String> addJar(String jarPath) { return null; }
    @Override public List<SinkClass> getSupportedSinks(SinkType sinkType, Collection<SinkClass> available) { return null; }

}

File: src/main/kotlin/moe/sota/decompiler/transformers/CFRTransformer.kt
package moe.sota.decompiler.transformers

import moe.sota.decompiler.models.FileModel
import org.benf.cfr.reader.api.CfrDriver
import org.benf.cfr.reader.api.ClassFileSource
import org.benf.cfr.reader.api.OutputSinkFactory
import org.benf.cfr.reader.api.OutputSinkFactory.Sink
import org.benf.cfr.reader.api.OutputSinkFactory.SinkClass
import org.benf.cfr.reader.api.OutputSinkFactory.SinkType
import org.benf.cfr.reader.bytecode.analysis.parse.utils.Pair
import org.benf.cfr.reader.util.getopt.OptionsImpl

class CFRTransformer : ITransformer, ClassFileSource, OutputSinkFactory, Sink<String> {
    private lateinit var fileModel: FileModel
    private lateinit var output: String

    override fun transform(fileModel: FileModel): String {
        this.fileModel = fileModel
        CfrDriver.Builder()
            .withClassFileSource(this)
            .withOptions(OPTIONS)
            .withOutputSink(this)
            .build()
            .analyse(listOf(fileModel.path))
        return if (output.startsWith("/")) output.substring(31) else output
    }

    override fun getClassFileContent(path: String): Pair<ByteArray, String> =
        Pair(fileModel.bytes, fileModel.path)

    override fun getPossiblyRenamedPath(path: String): String = path

    @Suppress("UNCHECKED_CAST")
    override fun <T> getSink(sinkType: SinkType, sinkClass: SinkClass): Sink<T> = this as Sink<T>

    override fun write(sinkable: String) {
        output = sinkable
    }

    override fun informAnalysisRelativePathDetail(usePath: String?, classFilePath: String?) {}

    override fun addJar(jarPath: String?): Collection<String>? = null

    override fun getSupportedSinks(
        sinkType: SinkType?,
        available: Collection<SinkClass>?,
    ): List<SinkClass>? = null

    companion object {
        private val OPTIONS =
            mapOf(
                OptionsImpl.DECOMPILE_INNER_CLASSES.name to "false",
                OptionsImpl.RELINK_CONSTANT_STRINGS.name to "false",
                OptionsImpl.REMOVE_INNER_CLASS_SYNTHETICS.name to "false",
                OptionsImpl.SHOW_CFR_VERSION.name to "false",
            )
    }
}
